algorithm: find the middle of linked-list

10kartik opened the pull request:
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not

* Fixes: #1095

appgurueu commented on Data-Structures/Linked-List/MiddleOfLinkedList.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * A LinkedList based solution for Detect a Cycle in a list
```
> Detect a Cycle in a list

Copy-paste mistake probably

appgurueu commented on Data-Structures/Linked-List/MiddleOfLinkedList.js:
```diff
@@ -0,0 +1,32 @@
+ * https://afteracademy.com/blog/middle-of-the-linked-list
+ */
+
+function main () {
```
This should be exported.

appgurueu commented on Data-Structures/Linked-List/MiddleOfLinkedList.js:
```diff
@@ -0,0 +1,32 @@
+
+    Link for the Problem: https://leetcode.com/problems/middle-of-the-linked-list/
+    */
+  const head = '' // Reference to head is given in the problem. So please ignore this line
```
The head should be a parameter.

appgurueu commented on Data-Structures/Linked-List/MiddleOfLinkedList.js:
```diff
@@ -0,0 +1,32 @@
+  return slow
+}
+
+main()
```
This will just crash.

appgurueu reviewed (changes requested):
This is not the form we expect algorithms to be in. Please:

* Add comments;
* Make this a re-usable and exported function;
* Add tests

10kartik commented:
Hello @appgurueu, I hope the algorithm is now as expected:
- Reusable and exported function solution
- Added tests
- Changed comments

Please review and suggest changes, if any.
Thanks!

appgurueu reviewed (changes requested):
There's no need for a class here; this should be a single function or a method of a linked list class (if there is a suitable one).

appgurueu commented on Data-Structures/Linked-List/MiddleOfLinkedList.js:
```diff
@@ -0,0 +1,27 @@
+    let fast = head
+    let slow = head
+
+    if (head.next == null) { return head }
```
This line is superfluous: If `head.next == null`, `fast.next` will be `null` in the first iteration, causing the `while` loop to not be entered at all. `slow`, which is returned, then is `head`.

10kartik replied:
Yeah, Right! Makes sense.

10kartik commented:
Hey @appgurueu, Adding a method in the existing LinkedList class made more sense.
Please review and suggest changes, if any.
Thanks!

appgurueu commented on Data-Structures/Linked-List/SinglyLinkedList.js:
```diff
@@ -193,6 +193,19 @@ class LinkedList {
   }
 
+  // Returns a reference to middle node of linked list
+  MiddleOfLL () {
```
please rename this to `findMiddle` or likewise

appgurueu commented on Data-Structures/Linked-List/test/SinglyLinkedList.test.js:
```diff
@@ -164,6 +164,32 @@ describe('SinglyLinkedList', () => {
+
+    let MiddleNodeOfLinkedList = list.MiddleOfLL(list.headNode)
+    // Middle node for list having single node
+    expect(MiddleNodeOfLinkedList.data).toEqual(1)
```
inline this to `expect(list.findMiddle()).toEqual(1)` (same below); you don't need to pass the `headNode` anymore

10kartik replied:
i think it should be `expect(list.findMiddle().data).toEqual(1)` cause findMiddle returns reference to the middle node rather than just its value.

appgurueu commented on Data-Structures/Linked-List/SinglyLinkedList.js:
```diff
@@ -193,6 +193,19 @@ class LinkedList {
+    let fast = this.headNode
+    let slow = this.headNode
+
+    while (fast != null && fast.next != null) {
```
Please use `!==` here.
